File: kikoapi/internal/tooling/registry.go
// Package tooling: Tool registry for AI (from kiko-api src/tooling).

package tooling

import "context"

// ToolDefinition is the schema for an AI tool (name, description, parameters).
type ToolDefinition struct {
	Name        string
	Description string
	Parameters  map[string]interface{}
	Required    []string
}

// ToolContext is passed to tool handlers (userId, chainId, etc.).
type ToolContext struct {
	UserID      string
	UserAddress string
	ChainID     int
}

// ToolHandler is the function type for executing a tool.
type ToolHandler func(ctx context.Context, args map[string]interface{}, tc *ToolContext) (interface{}, error)

// Tool is a registered tool with definition and handler.
type Tool struct {
	Definition ToolDefinition
	Handler    ToolHandler
}

// Registry holds tools by name.
type Registry struct {
	tools map[string]*Tool
}

// NewRegistry creates an empty registry.
func NewRegistry() *Registry {
	return &Registry{tools: make(map[string]*Tool)}
}

// Register adds a tool.
func (r *Registry) Register(tool *Tool) {
	if tool != nil {
		r.tools[tool.Definition.Name] = tool
	}
}

// GetTool returns the tool by name or nil.
func (r *Registry) GetTool(name string) *Tool {
	return r.tools[name]
}

// GetAllTools returns all registered tools.
func (r *Registry) GetAllTools() []*Tool {
	var out []*Tool
	for _, t := range r.tools {
		out = append(out, t)
	}
	return out
}

// GetDefinitions returns definitions for all tools.
func (r *Registry) GetDefinitions() []ToolDefinition {
	var out []ToolDefinition
	for _, t := range r.tools {
		out = append(out, t.Definition)
	}
	return out
}

// Execute runs a tool by name. Returns error if tool not found or handler errors.
func (r *Registry) Execute(ctx context.Context, name string, args map[string]interface{}, tc *ToolContext) (interface{}, error) {
	t := r.GetTool(name)
	if t == nil {
		return nil, ErrToolNotFound
	}
	return t.Handler(ctx, args, tc)
}

// ErrToolNotFound is returned when the tool name is not registered.
var ErrToolNotFound = &toolError{msg: "tool not found"}

type toolError struct{ msg string }

func (e *toolError) Error() string { return e.msg }

File: kiko-api/src/tooling/registry.ts
export const CHAIN_IDS = [1, 8453, 10, 42161, 137, 56, 900];
import { maybeRetryChainAwareToolExecution, prepareChainAwareToolExecution } from './chainAwareExecution.js';
import { ensureToolRegistryInitialized } from './bootstrap.js';

/**
 * Tool Definition Interfaces
 * 
 * Defines the structure for AI tools based on OpenAI/Claude function calling schemas.
 */

export interface ToolDefinition {
    name: string;
    description: string;
    parameters: {
        type: 'object';
        properties: Record<string, any>;
        required?: string[];
    };
    function?: {
        name: string;
        description: string;
        parameters: {
            type: 'object';
            properties: Record<string, any>;
            required: string[];
        };
    };
}

export interface ToolContext {
    userId?: string;
    userAddress?: string;
    chainId?: number;
    [key: string]: any;
}

export interface Tool<TArgs = any, TResult = any> {
    // The schema definition sent to the LLM
    definition: ToolDefinition;

    // The implementation function
    handler: (args: TArgs, context?: ToolContext) => Promise<TResult>;

    // Optional: Permissions level (e.g. 'read-only', 'auth-required', 'admin')
    permissions?: 'public' | 'user' | 'authenticated';
}

/**
 * Registry to hold all available tools
 */
export class ToolRegistry {
    private tools: Map<string, Tool> = new Map();
    private locks: Map<string, Promise<any>> = new Map();
    private static readonly NON_ISOLATED_TOOLS = new Set([
        'external_web_search',
        'get_token_info',
        'get_token_price',
        'get_historical_price',
        'get_trending_tokens',
        'get_early_buyers',
        'analyze_creator',
    ]);

    register(tool: Tool) {
        this.tools.set(tool.definition.name, tool);
    }

    getTool(name: string): Tool | undefined {
        ensureToolRegistryInitialized();
        return this.tools.get(name);
    }

    getAllTools(): Tool[] {
        ensureToolRegistryInitialized();
        return Array.from(this.tools.values());
    }

    getDefinitions(): ToolDefinition[] {
        return this.getAllTools()
            .map(t => t.definition)
            .sort((a, b) => a.name.localeCompare(b.name));
    }

    getAllDefinitions(): ToolDefinition[] {
        return this.getDefinitions();
    }

    async execute(name: string, args: any, context?: ToolContext): Promise<any> {
        ensureToolRegistryInitialized();
        const tool = this.getTool(name);
        if (!tool) {
            throw new Error(`Tool ${name} not found`);
        }
        if (ToolRegistry.NON_ISOLATED_TOOLS.has(name)) {
            const prepared = await prepareChainAwareToolExecution(name, args || {}, context);
            const executeOnce = (nextArgs: any) => tool.handler(nextArgs, context);
            const firstResult = await executeOnce(prepared.args);
            return maybeRetryChainAwareToolExecution(
                name,
                prepared.args,
                context,
                firstResult,
                prepared.meta,
                executeOnce,
            );
        }
        const key = this.buildIsolationKey(name, context);
        return this.runWithIsolation(key, async () => {
            const prepared = await prepareChainAwareToolExecution(name, args || {}, context);
            const executeOnce = (nextArgs: any) => tool.handler(nextArgs, context);
            const firstResult = await executeOnce(prepared.args);
            return maybeRetryChainAwareToolExecution(
                name,
                prepared.args,
                context,
                firstResult,
                prepared.meta,
                executeOnce,
            );
        });
    }

    private buildIsolationKey(name: string, context?: ToolContext): string {
        const userKey =
            context?.userId ||
            (context as any)?.privyDid ||
            context?.userAddress ||
            'anon';
        return `${name}:${userKey}`;
    }

    private async runWithIsolation<T>(key: string, fn: () => Promise<T>): Promise<T> {
        const prev = this.locks.get(key) || Promise.resolve();
        const startWait = Date.now();
        const current = prev.then(async () => {
            const waitMs = Date.now() - startWait;
            if (waitMs > 50) {
                console.log(`[ToolRegistry] Isolation wait ${waitMs}ms for ${key}`);
            }
            return fn();
        });
        this.locks.set(key, current);
        try {
            return await current;
        } finally {
            if (this.locks.get(key) === current) {
                this.locks.delete(key);
            }
        }
    }
}

export const toolRegistry = new ToolRegistry();
